iree/samples/simple_embedding/simple_embedding_run.c

```diff
@@ -0,0 +1,229 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A example of setting up the HAL mddule to run simple pointwise array
+// multiplication with the dylib driver.
+#include <stdio.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+
+#if IREE_ARCH_RISCV_64
+#include "iree/hal/dylib/registration/driver_module.h"
+#else
+#include "iree/hal/drivers/init.h"
+#endif
+
+#include "iree/modules/hal/hal_module.h"
+#include "iree/vm/api.h"
+#include "iree/vm/bytecode_module.h"
+
+// Compiled module embedded here to avoid file IO:
+#if IREE_ARCH_RISCV_64
+#include "iree/samples/simple_embedding/simple_embedding_test_llvm_aot_rv64.h"
+#else
+#include "iree/samples/simple_embedding/simple_embedding_test_bytecode_module_c.h"
+#endif
+
+iree_status_t Run(char* hal_driver_name) {
+  // TODO(benvanik): move to instance-based registration.
+  IREE_RETURN_IF_ERROR(iree_hal_module_register_types());
+
+  iree_vm_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance));
+
+#if IREE_ARCH_RISCV_64
+  // Only register dylib HAL driver
+  IREE_RETURN_IF_ERROR(iree_hal_dylib_driver_module_register(
+      iree_hal_driver_registry_default()));
+#else
+  // Register all drivers so it can be selected by the driver name.
+  IREE_RETURN_IF_ERROR(iree_hal_register_all_available_drivers(
+      iree_hal_driver_registry_default()));
+#endif
+
+  // Create the hal driver from the name. The driver name can be assigned as a
+  // hard-coded char array such as "dylib" as well.
+  iree_hal_driver_t* driver = NULL;
+  IREE_RETURN_IF_ERROR(iree_hal_driver_registry_try_create_by_name(
+      iree_hal_driver_registry_default(),
+      iree_make_cstring_view(hal_driver_name), iree_allocator_system(),
+      &driver));
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(iree_hal_driver_create_default_device(
+      driver, iree_allocator_system(), &device));
+  iree_vm_module_t* hal_module = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_hal_module_create(device, iree_allocator_system(), &hal_module));
+  iree_hal_driver_release(driver);
+
+  // Load bytecode module from the embedded data.
+#if IREE_ARCH_RISCV_64
+  const struct FileToc* module_file_toc =
+      simple_embedding_test_llvm_aot_rv64_create();
+#else
+  // Note the setup here only supports native build. The bytecode is not built
+  // for the cross-compile execution. The code can be compiled but it will
+  // hit runtime error in a cross-compile environment.
+  const struct FileToc* module_file_toc =
+      simple_embedding_test_bytecode_module_c_create();
+#endif
```

Comment:
`FileToc` was renamed to `iree_file_toc_t` in https://github.com/google/iree/pull/5578. Sync with `main` + apply this fix?

```suggestion
  // Load bytecode module from the embedded data.
#if IREE_ARCH_RISCV_64
  const struct iree_file_toc_t* module_file_toc =
      simple_embedding_test_llvm_aot_rv64_create();
#else
  // Note the setup here only supports native build. The bytecode is not built
  // for the cross-compile execution. The code can be compiled but it will
  // hit runtime error in a cross-compile environment.
  const struct iree_file_toc_t* module_file_toc =
      simple_embedding_test_bytecode_module_c_create();
#endif
```

Reply:
Done. Thanks for fixing the struct type.

Reply:
I do a similar thing to this here:
https://github.com/google/iree/blob/a8d76f0679e3d7fc6e7429355a53fcd55395deb7/iree/hal/local/elf/elf_module_test.cc#L32-L61
which is easier to scale out to new architectures (though still not great). I was going to add an IREE_ARCH_NAME define that would expand to one of these suffixes and then there's no need for any #ifdefs - just `"my_base_name_" IREE_ARCH_NAME ".vmfb"`

the real goal here is to get multi-architecture IREE modules so none of this is needed - that's next on my plate

Reply:
Understood. The challenge/difference in this example is that the default .vmfb is built from the iree_bytecode_module rule, so it does not have the arch_name suffix. I can change the "riscv_64" suffix for the riscv64 case, but it doesn't matter that much if we can't change the default.

Besides. In order to support the simple_embedding_test coverage, the default simple_embedding_bytecode_module.vmfb is built with both dylib and vulkan HAL, whereas for riscv we only going to test/support dylib. I want to distinguish that from the name too.

Agree supporting multi-architecture build will be better, however, it would only address the LLVM AOT path, however, not sure how/if it can be integrated with the Vulkan path.

Reply:
We'll be able to have all targets compiled in - and for a sample we'd do that - so a single vmfb for all CPU architectures and vulkan/cuda/etc - that just works with none of this #ifdef trickery. If we wanted something that only compiled in a single variant then we'd want to inject that instead of pulling it like we do here (or in the elf_module_test): the sample would just calls an extern C function (`load_sample_module()` etc) that then is controlled by the build system based on what you link in, which could just be whatever your CMake target is.

This is fine for now, though.